libs/internal/src/serialization/json_flag.cpp

```diff
@@ -0,0 +1,371 @@
+}
+
+tl::expected<std::optional<data_model::Flag>, JsonError> tag_invoke(
+    boost::json::value_to_tag<
+        tl::expected<std::optional<data_model::Flag>, JsonError>> const& unused,
+    boost::json::value const& json_value) {
+    boost::ignore_unused(unused);
+
+    REQUIRE_OBJECT(json_value);
+
+    auto const& obj = json_value.as_object();
+
+    data_model::Flag flag{};
+
+    PARSE_REQUIRED_FIELD(flag.key, obj, "key");
+
+    PARSE_CONDITIONAL_FIELD(flag.debugEventsUntilDate, obj,
+                            "debugEventsUntilDate");
+    PARSE_CONDITIONAL_FIELD(flag.salt, obj, "salt");
+    PARSE_CONDITIONAL_FIELD(flag.offVariation, obj, "offVariation");
+
+    PARSE_FIELD(flag.version, obj, "version");
+    PARSE_FIELD(flag.on, obj, "on");
+    PARSE_FIELD(flag.variations, obj, "variations");
+
+    PARSE_FIELD(flag.prerequisites, obj, "prerequisites");
+    PARSE_FIELD(flag.targets, obj, "targets");
+    PARSE_FIELD(flag.contextTargets, obj, "contextTargets");
+    PARSE_FIELD(flag.rules, obj, "rules");
+    PARSE_FIELD(flag.clientSide, obj, "clientSide");
+    PARSE_FIELD(flag.clientSideAvailability, obj, "clientSideAvailability");
+    PARSE_FIELD(flag.trackEvents, obj, "trackEvents");
+    PARSE_FIELD(flag.trackEventsFallthrough, obj, "trackEventsFallthrough");
+    PARSE_FIELD(flag.fallthrough, obj, "fallthrough");
+    return flag;
+}
+
+tl::expected<std::optional<data_model::Flag::VariationOrRollout>, JsonError>
+tag_invoke(boost::json::value_to_tag<
+               tl::expected<std::optional<data_model::Flag::VariationOrRollout>,
+                            JsonError>> const& unused,
+           boost::json::value const& json_value) {
+    boost::ignore_unused(unused);
+    REQUIRE_OBJECT(json_value);
+    auto const& obj = json_value.as_object();
+
+    std::optional<data_model::Flag::Rollout> rollout{};
+    PARSE_CONDITIONAL_FIELD(rollout, obj, "rollout");
+
+    if (rollout) {
+        return std::make_optional(*rollout);
+    }
+
+    std::optional<data_model::Flag::Variation> variation;
+
+    /* If there's no rollout, this must be a variation. If there's no variation,
+     * then this will be detected as a malformed flag at evaluation time. */
+    PARSE_CONDITIONAL_FIELD(variation, obj, "variation");
+
+    return variation;
+}
+
+namespace data_model {
+void tag_invoke(
+    boost::json::value_from_tag const& unused,
+    boost::json::value& json_value,
+    data_model::Flag::Rollout::WeightedVariation const& weighted_variation) {
+    auto& obj = json_value.emplace_object();
+    obj.emplace("variation", weighted_variation.variation);
+    obj.emplace("weight", weighted_variation.weight);
+
+    WriteMinimal(obj, "untracked", weighted_variation.untracked);
+}
+
+void tag_invoke(boost::json::value_from_tag const& unused,
+                boost::json::value& json_value,
+                data_model::Flag::Rollout::Kind const& kind) {
+    switch (kind) {
+        case Flag::Rollout::Kind::kUnrecognized:
```

Comment:
TODO

Reply:
Thinking about this, it feels like the long term solution shouldn't rely on "SDK deserializes data, then reserializes for persistence based on its in-memory model". Otherwise, old SDKs that aren't aware of new fields will drop them and "pollute" or otherwise downgrade the data. Instead, they should pass-through to the storage - save the CPU cycles and allow newer clients (including newer versions of the same SDK) to read the complete data.

Reply:
I agree with this conceptually, but it has a couple problems which I tried to reason through when we started the V2 process. We need to update individual records and we likely do not want to store or re-create the textual record in order to update. (double memory, or base storage in a non-typed form, like keeping the boost::json representation).

We get an init, we save that to the persistent store. 

We get a patch for a flag in that init, what do we do? Read back the json, deserialize it, do patch comparisons, update it? Ok, now how do we update the `allFlags` key? We cannot really patch JSON without putting into some representable form, changing it, and then serializing it. Basically we have parallel implementations of the patch update process.

I think that we possibly could do something like this, but it would require changing how we store the representation. I am not sure how yet. Certainly dropping the idea of the possibility of a store without the ability to scan, which is where the possibility of an 'all' key comes from.

Reply:
So, "all must be an aggregate of all the entries for the specified type". Which would hold true for dynamodb and redis.

Reply:
Added a todo link into the codebase for this. Deserves a more extended discussion, for sure.